read-fonts/src/tables/layout/closure.rs

```diff
@@ -0,0 +1,195 @@
+                feature_list
+                    .feature_records()
+                    .iter()
+                    .enumerate()
+                    .filter(|(_i, record)| features.contains(record.feature_tag()))
+                    .map(|(idx, _)| idx as u16)
+                    .collect()
+            }),
+            table_head,
+        }
+    }
+
+    /// Return true if the script limit has been exceeded or the script is visited before
+    pub(crate) fn script_visited(&mut self, s: &Script) -> bool {
+        if self.script_count > MAX_SCRIPTS {
+            return true;
+        }
+
+        self.script_count += 1;
+
+        let delta = (s.offset_data().as_bytes().as_ptr() as usize - self.table_head) as u32;
+        !self.visited_script.insert(delta)
+    }
+
+    /// Return true if the Langsys limit has been exceeded or the Langsys is visited before
+    pub(crate) fn langsys_visited(&mut self, langsys: &LangSys) -> bool {
+        if self.langsys_count > MAX_LANGSYS {
+            return true;
+        }
+
+        self.langsys_count += 1;
+
+        let delta = (langsys.offset_data().as_bytes().as_ptr() as usize - self.table_head) as u32;
+        !self.visited_langsys.insert(delta)
+    }
+
+    /// Returns true if the feature limit has been exceeded
+    pub(crate) fn feature_indices_limit_exceeded(&mut self, count: u16) -> bool {
+        let (new_count, overflow) = self.feature_index_count.overflowing_add(count);
+        if overflow {
+            self.feature_index_count = MAX_FEATURE_INDICES;
+            return true;
+        }
+        self.feature_index_count = new_count;
+        new_count > MAX_FEATURE_INDICES
+    }
+}
+
+impl ScriptList<'_> {
+    /// Return a set of all feature indices underneath the specified scripts, languages and features
+    ///
+    /// if no script is provided, all scripts will be queried
+    /// if no language is provided, all languages will be queried
+    /// if no feature is provided, all features will be queried
+    pub(crate) fn collect_features(
+        &self,
+        layout_table_head: usize,
+        feature_list: &FeatureList,
+        scripts: Option<&IntSet<Tag>>,
+        languages: Option<&IntSet<Tag>>,
+        features: Option<&IntSet<Tag>>,
+    ) -> Result<IntSet<u16>, ReadError> {
+        let mut out = IntSet::empty();
+        let mut c =
+            CollectFeaturesContext::new(features, layout_table_head, feature_list, &mut out);
+        let script_records = self.script_records();
+        let font_data = self.offset_data();
+        match scripts {
+            None => {
+                // All scripts
+                for record in script_records {
+                    let script = record.script(font_data)?;
+                    script.collect_features(&mut c, languages)?;
+                }
+            }
+            Some(scripts) => {
+                for idx in scripts.iter().filter_map(|tag| self.index_for_tag(tag)) {
```

Comment:
scripts might be inverted in which case this iteration will be over all integer values. If you remove the Option's as suggested above then you can change this to check is_inverted() and if it is inverted loop through script records filtering by the set and if not inverted use this approach.